### PR TITLE
Remove recursion in merkle tree implementation

### LINKED ISSUE #15: Remove recursion from `merkle tree` implementation.
MAM merkle tree currently uses recursion in `mam/v1/merkle.c` for traversal.
This makes debugging hard as no TCO happens

### PR DESCRIPTION
**Pull request by howjmay**
Remove the recursion in merkle tree in merkle.c

Fixes #15 

The memory depth is linealy propontional to the depth of merkle tree in the original recursive version of binary_tree_size(). but the memory dpeth is constant in the optimized version of binary_tree_size(). Moreover, the optimized version only calculates once instead of using iteration.

#### binary_tree_size()
The original binary_tree_size()'s stack memory size is "32 * tree_depth" bytes in which each frame size of binary_tress size() is 32 bytes.

The opitimized binary_tree_size()'s stack memory size is 64 bytes.

#### merkle_node_index_traverse()
The original merkle_node_index_traverse()'s stack memory size is "80 * tree_depth" bytes in which each frame size of merkle_node_index_traverse() is 50 bytes.

The opitimized merkle_node_index_traverse()'s stack memory size is 64 bytes.

### Performance 
Moreover, I replaced the function call, binary_tree_size(), in
merkle_node_index_traverse() with calculating directly.

#### binary_tree_size()
binary_tree_size() testing result for calculating a merkle tree
whose tree_depth is 20 and acc = 99 for 10000 times
=====================testing result==========================
    recursive one's cost clocks = 5414
    optimized one's cost clocks = 185

#### merkle_node_index_traverse()
merkle_node_index_traverse() testing result for calculating all the
nodes's indexes of a 20 level merkle tree
=====================testing result==========================
    recursive one's cost clocks = 318673
    optimized one's cost clocks = 185648

# Test Plan:
use test_merkle.c in mam/v1/tests

### CODE REVIEW

File: mam/v1/merkle.c [hunk 1]
```diff
@@ -28,7 +25,7 @@ size_t merkle_size(size_t const leaf_count) {
 
 size_t merkle_depth(size_t const node_count) {
   size_t depth = 0;
-
+
```
Comment (jserv) [inline mam/v1/merkle.c]: Avoid such non-functioned change.

Comment (howjmay) [reply mam/v1/merkle.c]: thanks. I fixed it

File: mam/v1/merkle.c [hunk 2]
```diff
@@ -28,7 +25,6 @@ size_t merkle_size(size_t const leaf_count) {
 
 size_t merkle_depth(size_t const node_count) {
   size_t depth = 0;
-
```
Comment (jserv) [inline mam/v1/merkle.c]: Don't change this line.

File: mam/v1/merkle.c [hunk 3]
```diff
@@ -38,15 +34,26 @@ size_t merkle_depth(size_t const node_count) {
...
-        1 + acc + binary_tree_size(0, tree_depth - 1), depth - 1,
-        width - (1 << (depth - 1)), tree_depth - 1);
+  if (!tree_depth)
+    return 0;
+
+  int depth_cursor = 1;
+  int tree_depth_width_divide_scale = 1;
+  int index = 0;
+  int width_cursor = width;
+  int tree_depth_width = 1 << depth;
```
Comment (jserv) [inline mam/v1/merkle.c]: `tree_depth_width` might not be a meaningful name. Can you improve it?

**Comment by jserv** [comment]
The commit messages should mention the following:
1. performance gain after switching to iteration-based implementation;
2. memory reduction ration because of no extra stack allocation;

Please use `git rebase -i` to rework message(s).

**Comment by jserv** [comment]
It would be explicit to change subject to `Remove recursion in merkle tree implementation".

**Comment by howjmay** [comment]
@thibault-martinez ?

**Comment by thibault-martinez** [comment]
Hi @HowJMay, thanks for this PR, I'll review it ASAP.
Great perf analysis by the way !

**Comment by howjmay** [comment]
thank you><

**Comment by howjmay** [comment]
@thibault-martinez Hi! 
It has been a long time since I sent this PR.
I wonder the status of code review of this PR.
I really hope that my PR can be merged into this porject. Thanks !

**Comment by thibault-martinez** [comment]
Hi @HowJMay , could you solve conflicts on this one please ?

**Comment by thibault-martinez** [comment]
Hi @HowJMay , could you solve conflict on this please ?

**Comment by thibault-martinez** [comment]
Hi @HowJMay , could you solve conflict on this please ?

**Comment by thibault-martinez** [comment]
Hi @HowJMay , could you solve conflict on this PR please ?

**Comment by thibault-martinez** [comment]
Sorry if you received my comment multiple times, GitHub is having difficulties today.
Your implementation of `merkle_node_index_traverse` gives wrong results for two reasons.
I'll point them out in the review with some other little things.

File: mam/v1/merkle.c [hunk 4]
```diff
@@ -38,15 +35,26 @@ size_t merkle_depth(size_t const node_count) {
 static size_t merkle_node_index_traverse(size_t const acc, size_t const depth,
```
Comment (thibault-martinez) [inline mam/v1/merkle.c]: You are not using `acc` at all.

Comment (thibault-martinez) [inline mam/v1/merkle.c]: You are not using `acc` at all.

File: mam/v1/merkle.c [hunk 5]
```diff
@@ -38,15 +35,26 @@ size_t merkle_depth(size_t const node_count) {
...
-                                      tree_depth - 1);
-  } else {
-    return merkle_node_index_traverse(
-        1 + acc + binary_tree_size(0, tree_depth - 1), depth - 1,
-        width - (1 << (depth - 1)), tree_depth - 1);
+  if (!tree_depth)
+    return 0;
+
+  int depth_cursor = 1;
+  int width_of_leaves_divide_scale = 1;
```
Comment (thibault-martinez) [inline mam/v1/merkle.c]: You don't need this variable, just use `depth_cursor`.

File: mam/v1/merkle.c [hunk 6]
```diff
@@ -38,15 +35,26 @@ size_t merkle_depth(size_t const node_count) {
...
+  int index = 0;
```
Comment (thibault-martinez) [inline mam/v1/merkle.c]: Initialize it at `depth + acc` instead of adding `depth` at the end.

File: mam/v1/merkle.c [hunk 7]
```diff
@@ -38,15 +35,26 @@ size_t merkle_depth(size_t const node_count) {
...
+  int width_cursor = width;
+  int width_of_leaves = 1 << depth;
+
+  while (depth_cursor <= depth) {
+    if (width_cursor >= (width_of_leaves >> width_of_leaves_divide_scale)) {
+      index += ((1 << (depth + 1)) - 1);
```
Comment (thibault-martinez) [inline mam/v1/merkle.c]: You are always adding the same tree here

File: mam/v1/merkle.c [hunk 8]
```diff
@@ -38,15 +35,26 @@ size_t merkle_depth(size_t const node_count) {
...
+    return 0;
+
+  int depth_cursor = 1;
+  int width_of_leaves_divide_scale = 1;
+  int index = 0;
+  int width_cursor = width;
+  int width_of_leaves = 1 << depth;
+
+  while (depth_cursor <= depth) {
+    if (width_cursor >= (width_of_leaves >> width_of_leaves_divide_scale)) {
```
Comment (thibault-martinez) [inline mam/v1/merkle.c]: Instead of shifting `1` to the left then to the right, just shift `1` to the left of `depth - depth_cursor`

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

**Comment by howjmay** [comment]
I am really sorry for that I did some really stupid error in the last version. 
I have fixed them out. Now, it can run successfully.

**Comment by thibault-martinez** [comment]
No worries, that's what reviews are for.
Just some final nits and it'll be good to merge.
Please solve conflicts as well.

File: mam/v1/merkle.c [hunk 9]
```diff
@@ -35,18 +32,33 @@ size_t merkle_depth(size_t const node_count) {
...
-  } else {
-    return merkle_node_index_traverse(
-        1 + acc + binary_tree_size(0, tree_depth - 1), depth - 1,
-        width - (1 << (depth - 1)), tree_depth - 1);
+  if (!tree_depth) return 0;
+
+  int depth_cursor = 1;
+  int index = depth + acc;
+  int width_cursor = width;
+  int width_of_leaves = 1 << depth;
```
Comment (thibault-martinez) [inline mam/v1/merkle.c]: Use `size_t` for these variables.

File: mam/v1/merkle.c [hunk 10]
```diff
@@ -35,18 +32,33 @@ size_t merkle_depth(size_t const node_count) {
...
                                          size_t const tree_depth) {
-  if (depth == 0) return acc;
-  if (width < (1 << (depth - 1))) {
-    return merkle_node_index_traverse(acc + 1, depth - 1, width,
-                                      tree_depth - 1);
-  } else {
-    return merkle_node_index_traverse(
-        1 + acc + binary_tree_size(0, tree_depth - 1), depth - 1,
-        width - (1 << (depth - 1)), tree_depth - 1);
+  if (!tree_depth) return 0;
```
Comment (thibault-martinez) [inline mam/v1/merkle.c]: Put braces around the `if`

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

**Review by thibault-martinez — APPROVED** [review]

**Comment by thibault-martinez** [comment]
@HowJMay ,
Thank you for this really great PR !
Please rebase to develop and it'll be good to merge.

**Comment by howjmay** [comment]
thank you for reviewing :)

**Comment by jserv** [comment]
@HowJMay, Don't use `git merge`. Instead, use `git rebase`: https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request

File: utils/merkle.c [hunk 11]
```diff
@@ -1,22 +1,12 @@
-/*
- * Copyright (c) 2018 IOTA Stiftung
- * https://github.com/iotaledger/entangled
- *
- * Refer to the LICENSE file for licensing information
- */
-
```
Comment (thibault-martinez) [inline utils/merkle.c]: Can you bring back the copyright ?

File: utils/merkle.c [hunk 12]
```diff
@@ -1,22 +1,12 @@
...
 #include <stdio.h>
 
 #include "common/sign/v2/iss_curl.h"
-#include "utils/merkle.h"
+#include "mam/v1/merkle.h"
```
Comment (thibault-martinez) [inline utils/merkle.c]: Wrong path

File: utils/merkle.c [hunk 13]
```diff
@@ -1,22 +1,12 @@
...
 
-static trit_t const merkle_null_hash[HASH_LENGTH_TRIT] = {0};
+static trit_t const merkle_null_hash[HASH_LENGTH] = {0};
```
Comment (thibault-martinez) [inline utils/merkle.c]: `HASH_LENGTH` does not exist

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

File: utils/merkle.c (hunk 13 above)
Comment (howjmay) [reply utils/merkle.c]: OK I will fix them this weekend sorry

**Review by thibault-martinez — APPROVED** [review]

**Comment by thibault-martinez** [comment]
@HowJMay Thanks again for this great PR !

**Pull request merged**